### PR TITLE
SAA-180 AM Events Endpoint: Client for (prison api) visits endpoint

### PR DESCRIPTION
**Pull request by markford-moj**
New client for prison api visits
Added prison api visits call to ScheduledEventService#getScheduledEventsByDateRange Added visits attribute to PrisonerScheduledEvents model Removed @Autowired annotation from ControllerAdvice - implicit constructor injection should make this annotation superflous.

### CODE REVIEW

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/ScheduledEventIntegrationTest.kt [hunk 1]
```diff
@@ -29,8 +30,10 @@ class ScheduledEventIntegrationTest : IntegrationTestBase() {
         dateRange.start,
         dateRange.endInclusive
       )
+    println(mapper.writeValueAsString(scheduledEvents))
```
Comment (matt-crowson-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/ScheduledEventIntegrationTest.kt:33]: Probably don't want/need this println ...

**Comment by matt-crowson-moj** [comment]
Are you going to add the ControllerTestBase class to address the comments on the previous PR?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ScheduledEventServiceTest.kt [hunk 2]
```diff
@@ -110,6 +131,54 @@ class ScheduledEventServiceTest {
...
+      )
+    ).thenReturn(schedVisitsMono)
+
+    whenever(
+      prisonApiClient.getScheduledCourtHearings(
+        900001, dateRange
+      )
+    ).thenReturn(courtHearingsMono)
+
+    Assertions.assertThatThrownBy {
```
Comment (matt-crowson-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ScheduledEventServiceTest.kt:146]: Minor.  Import tidy up on line 146.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/config/ControllerAdvice.kt [hunk 3]
```diff
@@ -24,7 +23,7 @@ import javax.persistence.EntityNotFoundException
 import javax.validation.ValidationException
 
 @RestControllerAdvice
-class ControllerAdvice(@Autowired private val mapper: ObjectMapper) : ResponseEntityExceptionHandler() {
+class ControllerAdvice(private val mapper: ObjectMapper) : ResponseEntityExceptionHandler() {
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/config/ControllerAdvice.kt:26]: Are you going to look at the tidying up the controller tests?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ScheduledEventService.kt [hunk 4]
```diff
@@ -22,15 +22,18 @@ class ScheduledEventService(private val prisonApiClient: PrisonApiClient) {
     }
     val prisonApiCalls = Mono.zip(
```
[the remark this answers is not in the document]
Comment (markford-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ScheduledEventService.kt:23]: Will bear in mind but im expecting to do two async calls for prisoner details and rollout prison in subsequent subtasks

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/ScheduledEventIntegrationTest.kt (hunk 1 above)
Comment (markford-moj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/ScheduledEventIntegrationTest.kt:33]: Damn just had that for my mapping doc

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/config/ControllerAdvice.kt (hunk 3 above)
Comment (markford-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/config/ControllerAdvice.kt:26]: Wasn't sure if you already had a PR for the test base class Matt. Tbh test base classes are a bit of a pet dislike of mine. Happy to use it tho if you definitely want to go down this road.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ScheduledEventServiceTest.kt [hunk 5]
```diff
@@ -76,13 +89,19 @@ class ScheduledEventServiceTest {
       )
     ).thenReturn(schedAppointmentsMono)
 
+    whenever(
```
Comment (matt-crowson-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ScheduledEventServiceTest.kt:92]: A few of the fixtures aren't actually need in this test `getScheduledEventsByDateRange - prisoner details error` lines 86 to 102.  Sorry only just noticed this one.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ScheduledEventServiceTest.kt [hunk 6]
```diff
@@ -97,6 +116,50 @@ class ScheduledEventServiceTest {
...
+      prisonApiClient.getPrisonerDetails("A11111A")
+    ).thenReturn(prisonerDetailsMono)
+
+    whenever(
+      prisonApiClient.getScheduledAppointments(
+        900001, dateRange
+      )
+    ).thenReturn(schedAppointmentsMono)
+
+    whenever(
```
Comment (matt-crowson-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ScheduledEventServiceTest.kt:134]: Lines 134-144 not needed.  Taking them out has no impact on the test.

**Review by matt-crowson-moj — APPROVED** [review]
LGTM a couple of minors added that I didn't spot last time.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ScheduledEventServiceTest.kt (hunk 5 above)
Comment (markford-moj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ScheduledEventServiceTest.kt:92]: Good sport - will sort in next subtask

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ScheduledEventServiceTest.kt (hunk 6 above)
Comment (markford-moj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ScheduledEventServiceTest.kt:134]: Not sure we can guarantee that due to the async behaviour

**Pull request merged**